Add DeepSeek support

Cupnfish opened the pull request:
- Added support for DeepSeek as a new language model provider in Zed Assistant
- Implemented streaming API support for real-time responses from DeepSeek models.
- Added a configuration UI for DeepSeek API key management and settings.
- Updated documentation with detailed setup instructions for DeepSeek integration.
- Added DeepSeek-specific icons and model definitions for seamless integration into the Zed UI.
- Integrated DeepSeek into the language model registry, making it available alongside other providers like OpenAI and Anthropic.

Release Notes:

- Added support for DeepSeek to the Assistant.

Cupnfish committed:
feat: add DeepSeek integration for Zed Assistant

- Add DeepSeek as a new language model provider
- Implement DeepSeek API client with streaming support
- Add DeepSeek configuration UI and settings
- Update documentation with DeepSeek setup instructions
- Add DeepSeek icon and model definitions
- Integrate DeepSeek into language model registry

Cupnfish committed:
feat: add DeepSeek AI icon to assets/icons directory

maxdeviant commented:
> deepseek crate is missing a license.
> 
> ```shell
> cd crates/deepseek
> ln -s ../../LICENSE-GPL
> ```
> 
> In the future if you use a non-main branch on your fork and select "Allow Maintainers to Make changes to my branch" we can contribute fixes.

It looks like we already can push to it:

<img width="319" alt="Screenshot 2025-01-23 at 6 41 13 PM" src="https://github.com/user-attachments/assets/e2830388-28d9-4ad6-9127-d700f4c042d9" />

Cupnfish committed:
docs: add symlink to GPL license in deepseek crate

Cupnfish committed:
fix: resolve consecutive messages error for DeepSeek Reasoner

- Add message merging to fix '400 Bad Request' error for Reasoner model
- Address error: 'deepseek-reasoner does not support successive user or assistant messages'
- Merge consecutive messages of same role before sending to API
- Handle User, System, Assistant and Tool message types
- Preserve tool_calls in merged assistant messages
- Add comprehensive test coverage for message merging logic
- Optimize string concatenation with reserve and push_str
- Make Model::id const for better performance

Fixes ERROR assistant_context_editor] Failed to connect to DeepSeek API: 400 Bad Request {'error':{'message':'deepseek-reasoner does not support successive user or assistant messages (messages[0] and messages[1] in your input). You should interleave the user/assistant messages in the message sequence.','type':'invalid_request_error','param':null,'code':'invalid_request_error'}}

Cupnfish commented:
https://github.com/deepseek-ai/DeepSeek-R1/issues/21

Fixed issues:

```
ERROR assistant_context_editor] Failed to connect to DeepSeek API: 400 Bad Request {"error":{"message":"deepseek-reasoner does not support successive user or assistant messages (messages[0] and messages[1] in your input). You should interleave the user/assistant messages in the message sequence.","type":"invalid_request_error","param":null,"code":"invalid_request_error"}}
```

However, I'm not sure if the chosen method is the most appropriate. Welcome to discuss

Cupnfish committed:
refactor: simplify error handling and message merging logic in deepseek API

- Removed custom DeepSeekError struct and simplified error handling in the `complete` function.
- Refactored `merge_consecutive_messages` to use a more straightforward iteration approach.
- Introduced helper functions `merge_messages` and `can_merge` to improve readability.
- Updated tests to reflect changes in message merging behavior.

Cupnfish committed:
feat: add support for custom models in DeepSeek provider

- Added `Custom` variant to `deepseek::Model` enum to support custom models
- Updated `id()`, `display_name()`, `max_token_count()`, and `max_output_tokens()` methods to handle custom models
- Modified `provided_models()` in DeepSeek provider to include custom models from settings

Cupnfish committed:
refactor: move message merging logic to request conversion and remove unused complete function

- Removed the `complete` function and its associated tests from `deepseek.rs`
- Moved message merging logic from `deepseek.rs` to `request.rs`'s `into_deepseek` method
- Simplified `stream_completion` by removing redundant message merging
- Updated message merging to handle both user and assistant messages in one pass
- Maintained existing functionality while improving code organization

Cupnfish commented:
@shishkin Sorry for the misunderstanding, deepseek does indeed support code completion, but it was not implemented at the time. I have already revised the release notes. If anyone is interested in implementing this feature, they can refer to this document:

https://api-docs.deepseek.com/guides/fim_completion
https://api-docs.deepseek.com/guides/chat_prefix_completion

akilcelen commented:
hi,
Deepseek Gets cut off mid-work. Why is that? I can never complete my codes because of that...

guilhermerodz commented:
I've encountered an issue with AI integration in other text editors where Deepseek would cut off mid-work. However, this doesn't happen with LM Studio in Zed (since it's local). I wonder if this update will address edge cases when Deepseek handles complex work, which often takes longer, to prevent cuts.

Just like @akilcelen mentioned
> hi, Deepseek Gets cut off mid-work. Why is that? I can never complete my codes because of that...

savoygrizzly commented:
When is this issue set to release ? R1 is a game changer, and I dont yet have a project digits to run it on

maxdeviant committed:
Merge branch 'main' into pr/23551

maxdeviant committed:
Update to match GPUI changes

maxdeviant committed:
Change color for icon

maxdeviant reviewed (approved):
Thank you!
